source/Fleet.cpp

```diff
@@ -558,49 +541,13 @@ void Fleet::Place(const System &system, Ship &ship)
 
 int64_t Fleet::Strength() const
 {
-	if(!total || variants.empty())
+	if(variants.empty())
 		return 0;
 
 	int64_t sum = 0;
-	for(const Variant &variant : variants)
-	{
-		int64_t thisSum = 0;
-		for(const Ship *ship : variant.ships)
-			thisSum += ship->Cost();
-		sum += thisSum * variant.weight;
-	}
-	return sum / total;
-}
-
-
-
-Fleet::Variant::Variant(const DataNode &node)
-{
-	weight = 1;
-	if(node.Token(0) == "variant" && node.Size() >= 2)
-		weight = node.Value(1);
-	else if(node.Token(0) == "add" && node.Size() >= 3)
-		weight = node.Value(2);
-
-	for(const DataNode &child : node)
-	{
-		int n = 1;
-		if(child.Size() >= 2 && child.Value(1) >= 1.)
-			n = child.Value(1);
-		ships.insert(ships.end(), n, GameData::Ships().Get(child.Token(0)));
-	}
-}
-
-
-
-const Fleet::Variant &Fleet::ChooseVariant() const
-{
-	// Pick a random variant based on the weights.
-	unsigned index = 0;
-	for(int choice = Random::Int(total); choice >= variants[index].weight; ++index)
-		choice -= variants[index].weight;
-
-	return variants[index];
+	for(const auto &variant : variants)
+		sum += variant.Strength() * variant.Weight();
+	return sum / variants.TotalWeight();
```

Comment:
This could be replaced with the hypothetical `Average` method I mentioned:
```c++
return variants.Average(std::mem_fn(&Variant::Strength));

template <class Type>
decltype(.....) WeightedList::Average(std::mem_fn<Type> fn) const
{
  fn::result_type acc{};
  for(auto it = choices.begin(); it != choices.end(); ++it)
    acc += fn(it) * it->Weight();
  return acc / TotalWeight(); // maybe cast `acc` to double for the division?
}
```

(There's some template fun needed to declare the signature of the `WeightedList::Average` method so that it will work for arbitrary T::* methods, but it's doable.)

Reply:
But then you'll make me write test cases for this. :(

Reply:
Yes, making sure code works the way it was intended to, automatically, even with future changes, is a godsend. (It's even better when you write/update the test for the change you are going to make before you do it, because then you know the test works, too.)

Reply:
Well I'm not even managing to get this working anyway.

Reply:
I added it for you